crates/matrix-sdk/src/notification_settings.rs

```diff
@@ -0,0 +1,1116 @@
+//! High-level push notification settings API
+
+use ruma::{
+    api::client::push::{delete_pushrule, set_pushrule, set_pushrule_enabled, RuleScope},
+    push::{
+        Action, NewConditionalPushRule, NewPushRule, NewSimplePushRule, PredefinedContentRuleId,
+        PredefinedOverrideRuleId, PredefinedUnderrideRuleId, PushCondition, RuleKind, Ruleset,
+        Tweak,
+    },
+    RoomId,
+};
+
+use crate::{error::NotificationSettingsError, Client, Result};
+
+/// Enum representing the push notification modes for a room.
+#[derive(Debug, Clone, PartialEq)]
+pub enum RoomNotificationMode {
+    /// Receive notifications for all messages.
+    AllMessages,
+    /// Receive notifications for mentions and keywords only.
+    MentionsAndKeywordsOnly,
+    /// Do not receive any notifications.
+    Mute,
+}
+
+/// A high-level API to manage the client owner's push notification settings.
+#[derive(Debug, Clone)]
+pub struct NotificationSettings {
```

Comment:
(This is the large change I'm mentioning in the main review comment.)

This object mixes a few responsibilities:

- computing the actual rules from observing the raw Ruma `Ruleset`
- interacting with the server to add/remove things

The former tends to not even make use of `self`, in methods like `get_custom_rules_for_room`, `get_user_defined_room_notification_mode`, and a few others. This suggests a different implementation:

- create a wrapper struct for `Ruleset`, which only job is to interact with the inner `Ruleset`.
- then have this `NotificationSettings` contain this wrapper (I don't think it should be the FFI's responsbility to own the `Ruleset`), and expose extra methods to interact with the server and update the ruleset.

That would blackbox the different problems, and if the `Ruleset`'s wrapper needs to interact with the server, it can use a command pattern and return a vec of commands to be applied by the client thereafter. Would probably help with testing, and also this would give us the possibility to split this PR into smaller ones that are easier to implement, test, review and merge :grin: : one PR for the `Ruleset` wrapper and its tests, one for the `NotificationSettings`. How does that sound?